subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -459,6 +460,8 @@ public void execute(BuildOperationContext buildOperationContext) {
                 dependencyResolutionListeners.getSource().afterResolve(incoming);
                 // Discard listeners
                 dependencyResolutionListeners.removeAll();
+
+                lockAttributes();
```

Comment:
Makes sense. I moved it up to the beginning of the _resolve_ code block.

Note that lockAttributes is already called earlier in a resolution process here:
https://github.com/gradle/gradle/blob/7eb9d8b17d1e8e0e16d945dbc4300137dca194c6/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/DefaultConfigurationComponentMetaDataBuilder.java#L47-L47

I added this to make sure a lock is performed when `resolve()` is called on a configuration from another place. I was triggered to do this by this test:
https://github.com/gradle/gradle/blob/f61eb3671920b388f505b7c5ebacc4b95f99457b/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/UnsupportedConfigurationMutationTest.groovy#L455-L455